llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp

```diff
@@ -2878,7 +2878,8 @@ Instruction *InstCombinerImpl::visitFNeg(UnaryOperator &I) {
 
     // -(Cond ? X : C) --> Cond ? -X : -C
     // -(Cond ? C : Y) --> Cond ? -C : -Y
-    if (match(X, m_ImmConstant()) || match(Y, m_ImmConstant())) {
+    if ((match(X, m_ImmConstant()) && !isa<ScalableVectorType>(X->getType())) ||
```

Comment:
We can use `ConstantFoldUnaryOpOperand(Instruction::FNeg, C, DL)` instead of using IRBuilder to fix this issue.

if (match(X, m_ImmConstant(XC)) || match(Y, m_ImmConstant(YC))) {
   Value *NegX = ConstantFoldUnaryOpOperand(Instruction::FNeg, XC, DL);
   Value *NegY = ConstantFoldUnaryOpOperand(Instruction::FNeg, YC, DL);
   if (NegX || NegY) {
      if (!NegX) NegX = Builder.CreateFNeg...;
      if (!NegY) NegY = Builder.CreateFNeg...;
      SelectInst *NewSel = SelectInst::Create(Cond, NegX, NegY);
      propagateSelectFMF(NewSel, /*CommonOperand=*/true);
      return NewSel;
   }
}

Reply:
Thanks for having a look, Yingwei.  I'll test and upload the revised version.

Reply:
The IRBuilder should do this like in other cases?